File: backends/node/shared/implementations/caching/distributed/redis/src/handlers/r/exists.ts
import type Redis from "ioredis";
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import { redisErrorResult } from "../../redis-error-result.js";
import type { DistributedCache } from "@d2/interfaces";

type Input = DistributedCache.ExistsInput;
type Output = DistributedCache.ExistsOutput;

export class Exists extends BaseHandler<Input, Output> implements DistributedCache.IExistsHandler {
  private readonly redis: Redis;

  constructor(redis: Redis, context: IHandlerContext) {
    super(context);
    this.redis = redis;
  }

  protected async executeAsync(input: Input): Promise<D2Result<Output | undefined>> {
    try {
      const count = await this.redis.exists(input.key);
      return D2Result.ok({ data: { exists: count > 0 } });
    } catch (err: unknown) {
      return redisErrorResult(err);
    }
  }
}

File: backends/dotnet/shared/Implementations/Caching/Distributed/DistributedCache.Redis/Handlers/R/Exists.cs
// -----------------------------------------------------------------------
// <copyright file="Exists.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.DistributedCache.Redis.Handlers.R;

using D2.Shared.Handler;
using D2.Shared.Result;
using Microsoft.Extensions.Logging;
using StackExchange.Redis;
using H = D2.Shared.Interfaces.Caching.Distributed.Handlers.R.IRead.IExistsHandler;
using I = D2.Shared.Interfaces.Caching.Distributed.Handlers.R.IRead.ExistsInput;
using O = D2.Shared.Interfaces.Caching.Distributed.Handlers.R.IRead.ExistsOutput;

/// <summary>
/// Handler for checking if a key exists in the Redis distributed cache.
/// </summary>
public partial class Exists : BaseHandler<H, I, O>, H
{
    private readonly IConnectionMultiplexer r_redis;

    /// <summary>
    /// Initializes a new instance of the <see cref="Exists"/> class.
    /// </summary>
    ///
    /// <param name="redis">
    /// The Redis connection multiplexer.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public Exists(
        IConnectionMultiplexer redis,
        IHandlerContext context)
        : base(context)
    {
        r_redis = redis;
    }

    /// <inheritdoc/>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        try
        {
            // Connect to Redis and check if the key exists.
            var db = r_redis.GetDatabase();
            var exists = await db.KeyExistsAsync(input.Key);

            // Return the result.
            return D2Result<O?>.Ok(
                new O(exists));
        }
        catch (RedisException ex)
        {
            LogExistsFailed(Context.Logger, ex, input.Key, TraceId);

            return D2Result<O?>.ServiceUnavailable(traceId: TraceId);
        }

        // Let the base handler catch any other exceptions.
    }

    /// <summary>
    /// Logs that a Redis exception occurred while checking key existence.
    /// </summary>
    [LoggerMessage(EventId = 1, Level = LogLevel.Error, Message = "RedisException occurred while checking existence of key '{Key}'. TraceId: {TraceId}")]
    private static partial void LogExistsFailed(ILogger logger, Exception ex, string key, string? traceId);
}
